src/server/server.ts

```diff
@@ -200,6 +259,93 @@ registerRoute('post', '/compile/snippet', compileCommon);
  */
 registerRoute('post', '/compile/page', (req, res) => compileCommon(req, res, true /*wrapWithRunnerChrome*/));
 
+registerRoute('get', '/open-in-playground/:correlationId/:host/:type/:id/:filename', async (req, res) => {
+    let relativePath, templateName;
+    switch (req.params.host.toUpperCase()) {
+        case 'EXCEL':
+            relativePath = 'xl';
+            templateName = 'excel-template';
+            break;
+        case 'WORD':
+            relativePath = 'word';
+            templateName = 'word-template';
+            break;
+        case 'POWERPOINT':
+            relativePath = 'ppt';
+            templateName = 'powerpoint-template';
+            break;
+        default:
+            throw new Error(`Unsupported host: ${req.params.host}`);
+    }
+
+    const correlationId = req.params.correlationId;
+    let directoryInfo = generatedDirectories[correlationId];
+    // Check if file already exists on server for correlation id
+    if (directoryInfo) {
+        directoryInfo.isBeingRead = true;
+        fs.createReadStream(directoryInfo.name).pipe(res);
+        res.attachment(req.params.filename);
+        res.on('finish', () => {
+            directoryInfo.isBeingRead = false;
+        });
+    } else {
+        return getVersionNumber()
+            .then(versionNumber => {
+                let timestamp = (new Date()).getTime();
+                let relativeFilePath = `working/${templateName}${timestamp}`;
+                let extractDirName = path.resolve(__dirname, relativeFilePath);
+                let zip = fs.createReadStream(path.resolve(__dirname, `${templateName}.zip`)).pipe(unzip.Extract({ path: extractDirName }));
+
+                zip.on('close', () => {
+                    let xmlFileName = `${extractDirName}/${relativePath}/webextensions/webextension1.xml`;
+                    (new Promise<string>((resolve, reject) => {
+                        fs.readFile(xmlFileName, (err, data) => {
```

Comment:
I would create readFileAsPromise as a helper method outside of this function, then we can reuse it in many places.